Com 1245

tresorluz opened the pull request.

tresorluz committed:
COM-1245 Added new fields in the Patient Information report

tresorluz committed:
COM-1245 Added a new function in the patient information report

tresorluz committed:
COM-1245 Added new fields into the template

tresorluz committed:
COM-1245 Added new functionality to combine 2 drugs in the current regiment column

tresorluz committed:
Added unit test to the patient information line listing report

tmvumbi2 commented on openmrs/apps/reports/GEORGETOWN_PATIENT_INFORMATION_REPORT/sql/georgetownPatientInformationReport.sql:
```diff
@@ -29,7 +33,12 @@ SELECT
-    getProgramAttributeValueWithinReportingPeriod(p.patient_id, "#startDate#", "#endDate#", "8bb0bdc0-aaf3-4501-8954-d1b17226075b", "HIV_PROGRAM_KEY") as "APS Name"
+    getProgramAttributeValueWithinReportingPeriod(p.patient_id, "#startDate#", "#endDate#", "8bb0bdc0-aaf3-4501-8954-d1b17226075b", "HIV_PROGRAM_KEY") as "APS Name",
+    getObsCodedValue(p.patient_id, "76a912af-702e-47a0-83c4-246900861296") as "disclosureStatus",
+    getObsCodedValue(p.patient_id, "f0447183-d13f-463d-ad0f-1f45b99d97cc") as "tbScreening",
```
@tresorluz this concept also exists in the ANC Initial form. Let's rather use the function `getObsCodedValueInSectionByNames` to ensure we are retrieving only the value coming from the TB Form

tmvumbi2 commented on metadata/reportssql/drug_functions.sql:
```diff
@@ -694,6 +694,69 @@ BEGIN
+        AND o.scheduled_date
+        AND d.name LIKE "INH%"
+        AND drugOrderIsDispensed(p_patientId, o.order_id)
+    ORDER BY o.scheduled_date ASC
```
@tresorluz to have the most recent date on top, this should be a descending order  (i.e. `DESC`)

tmvumbi2 commented on metadata/reportssql/drug_functions.sql:
```diff
@@ -694,6 +694,69 @@ BEGIN
+        JOIN orders o ON o.order_id = do.order_id  AND o.voided = 0
+        JOIN drug d ON d.drug_id = do.drug_inventory_id AND d.retired = 0
+    WHERE o.patient_id = p_patientId
+        AND o.scheduled_date
```
@tresorluz this condition looks incomplete

tmvumbi2 commented on metadata/reportssql/drug_functions.sql:
```diff
@@ -694,6 +694,69 @@ BEGIN
+
+DELIMITER $$
+CREATE FUNCTION getINHStartDate(
+    p_patientId INT(11)
```
@tresorluz could we confirm with JP if we should be including the most recent INH that was started within the reporting period?

tresorluz replied:
Yeah we take the most recent date within the reporting period

tmvumbi2 commented on metadata/reportssql/drug_functions.sql:
```diff
@@ -694,6 +694,69 @@ BEGIN
+    DECLARE result INT(11);
+
+    SELECT 
+        (CASE
```
@tresorluz we can re-use the function `calculateDurationInMonths` here, it handles all the duration units already (there are 5 units available)

tmvumbi2 commented on metadata/reportssql/drug_functions.sql:
```diff
@@ -694,6 +694,69 @@ BEGIN
+        AND o.scheduled_date
+        AND d.name LIKE "INH%"
+        AND drugOrderIsDispensed(p_patientId, o.order_id)
+    ORDER BY o.scheduled_date ASC
```
@tresorluz to have the most recent date on top, this should be a descending order (i.e. DESC)

tmvumbi2 commented on metadata/reportssql/drug_functions.sql:
```diff
@@ -694,6 +694,69 @@ BEGIN
+        JOIN orders o ON o.order_id = do.order_id  AND o.voided = 0
+        JOIN drug d ON d.drug_id = do.drug_inventory_id AND d.retired = 0
+    WHERE o.patient_id = p_patientId
+        AND o.scheduled_date
```
@tresorluz this condition looks incomplete as well

tmvumbi2 commented on metadata/reportssql/drug_functions.sql:
```diff
@@ -694,6 +694,69 @@ BEGIN
+
+DELIMITER $$
+CREATE FUNCTION getINHDuration(
+    p_patientId INT(11)
```
@tresorluz here as well, could we confirm with JP if we should be including the duration of the most recent INH that was started within the reporting period?

tmvumbi2 commented on metadata/reportssql/drug_functions.sql:
```diff
@@ -894,6 +957,43 @@ BEGIN
+                do.duration,
+                c.uuid)
+            )
+    ORDER BY calculateDurationInDays(o.scheduled_date,do.duration,c.uuid) ASC,
```
@tresorluz we won't need `ORDER BY` in this case, the result is being grouped and concatenated. Also, removing this will help improve the report performance because ORDER BY calling a function is an expensive operation

tmvumbi2 commented on openmrs/apps/reports/GEORGETOWN_PATIENT_INFORMATION_REPORT/sql/georgetownPatientInformationReport.sql:
```diff
@@ -9,11 +9,15 @@ SELECT
     getFirstARVPrescribed(p.patient_id) as "regimentAtArtInitiation",
     getARVTherapeuticLineAtInitiation(p.patient_id) as "lineAtInitiation",
-    getActiveARVWithLowestDispensationPeriod(p.patient_id, "2000-01-01", "2100-01-01") as "currentRegimen",
+    concatenateARTDrugs(p.patient_id, "2000-01-01", "2100-01-01")as "currentRegimen",
```
@tresorluz If JP agrees with the definition I asked about in the slack chat, we'll need to plug in here the report start date and end date variables. I have included the comment down here too for easy reference:

is the following definition of “Current ARV Drugs” correct?
```ARV drugs dispensed to patients include those with treatments starting within the reporting period, as well as treatments starting prior to the reporting period BUT ending either during or after the reporting period. ARV drugs that started after the reporting period will not be included. Also, ARV that both started and ended before the reporting period will be excluded.```

tresorluz replied:
JP didn't agree with this

tmvumbi2 commented on report-testing-framework/report-testing/src/test/java/org/jembi/bahmni/report_testing/georgetown_reports/GeorgetownPatientInformationReportTests.java:
@tresorluz we also need to update the VBA script that copy data from the hidden sheet "report" into the main sheet displayed, to include the new columns that have been added. The VBA function name is `CopyReport_to_PasteSheet1` (look at the code below the comment `'Copy Data`. Please ping me if you need help with this edit

tmvumbi2 reviewed (changes requested):
Thanks @tresorluz , that was a lot of work completed! I left a few feedback and fixes to wrap this up

tresorluz committed:
COM-1245 Added new columns to the patient informaton templates

tresorluz committed:
COM-1245 Attended to code review feedbacks

tmvumbi2 committed:
COM-1245 Fix template - include last column in copy range
The last column (column AO) wasn't included in the range of cells being copied into the visible sheet; this commit fixes the issue
